src/lara-plugin/plugins/plugin-context.ts

```diff
@@ -0,0 +1,218 @@
+
+/** Data generated by LARA, passed to .initPlugin function, and then used to generate final IPluginRuntimeContext */
+export interface IPluginRuntimeContextOptions extends IPluginCommonOptions {
+  /** Type of the plugin */
+  type: "runtime";
+  /** The saved learner data for this instance (if available). */
+  learnerState: string | null;
+  /** URL used to save plugin state. */
+  learnerStateSaveUrl: string;
+  /** The run ID for the current LARA run. */
+  runId: number;
+  /** The portal remote endpoint (if available). */
+  remoteEndpoint: string | null;
+  /** The current users email address (if available). */
+  userEmail: string | null;
+  /** The portal URL for class details (if available). */
+  classInfoUrl: string | null;
+  /** The ID of the Embeddable that has been added to page, this embeddable refers to the plugin instance */
+  embeddablePluginId: number | null;
+  /** URL of the resource associated with the current run (sequence URL or activity URL) */
+  resourceUrl: string;
+}
+
+export interface IPluginAuthoringContextOptions extends IPluginCommonOptions {
+  /** Type of the plugin */
+  type: "authoring";
+  /** URL used to save plugin authoring state. */
+  authorDataSaveUrl: string;
+}
+
+/** Data generated by LARA, passed to .initPlugin function, and then used to generate final IEmbeddableRuntimeContext */
+export interface IEmbeddableContextOptions {
+  /** Embeddable container. */
+  container: HTMLElement;
+  /****************************************************************************
+   Serialized form of the embeddable. Defined by LARA export code, so it's format cannot be specified here.
+   Example (interactive):
+   ```
+   {
+     aspect_ratio_method: "DEFAULT",
+     authored_state: null,
+     click_to_play: false,
+     enable_learner_state: true,
+     name: "Test Interactive",
+     native_height: 435,
+     native_width: 576,
+     url: "http://concord-consortium.github.io/lara-interactive-api/iframe.html",
+     type: "MwInteractive",
+     ref_id: "86-MwInteractive"
+   }
+   ```
+   ****************************************************************************/
+  laraJson: any;
+  /** Interactive state URL, available only when plugin is wrapping an interactive. */
+  interactiveStateUrl: string | null;
+  /** True if the interactive is immediately available for use */
+  interactiveAvailable: boolean;
+}
+
+const ajaxPromise = (url: string, data: any): Promise<string> => {
+  return new Promise((resolve, reject) => {
+    $.ajax({
+      url,
+      type: "PUT",
+      data,
+      success(result) {
+        resolve(result);
+      },
+      error(jqXHR, errText, err) {
+        reject(err);
+      }
+    });
+  });
+};
+
+export const saveLearnerPluginState = (learnerStateSaveUrl: string, state: string): Promise<string> => {
+  return ajaxPromise(learnerStateSaveUrl, { state });
+};
+
+export const saveAuthoredPluginState = (authoringSaveStateUrl: string, authorData: string): Promise<string> => {
```

Comment:
Right.  My approach, at present, was to bring in entire modules when the module contained a dependency that was needed.  I was trying to change modules as little as possible for now  (for the most part, module changes were limited to export, import and lint changes).  My thought was we could strip out unused functions when all of the plugin work was complete (rather than trying to wrestle with following dependency chains and bring in function, types, etc. one by one).